runtime/src/mips.rs

```diff
@@ -0,0 +1,721 @@
+        /// proceed to ratification process.
+        fn on_initialize(n: T::BlockNumber) {
+            if let Err(e) = Self::end_block(n) {
+                sr_primitives::print(e);
+            }
+        }
+
+    }
+}
+
+impl<T: Trait> Module<T> {
+    /// Retrieve all proposals that need to be closed as of block `n`.
+    pub fn proposals_maturing_at(n: T::BlockNumber) -> Vec<(MipsIndex, T::Hash)> {
+        Self::proposal_meta()
+            .into_iter()
+            .filter(|meta| meta.end == n)
+            .map(|meta| (meta.index, meta.proposal_hash))
+            .collect()
+    }
+
+    // Private functions
+
+    /// Runs the following procedure:
+    /// 1. Find all proposals that need to end as of this block and close voting
+    /// 2. Tally votes
+    /// 3. Submit any proposals that meet the quorum threshold, to the governance committee
+    fn end_block(block_number: T::BlockNumber) -> Result {
+        sr_primitives::print("end_block");
+
+        // Find all matured proposals...
+        for (index, hash) in Self::proposals_maturing_at(block_number).into_iter() {
+            // Tally votes and create referendums
+            Self::tally_votes(index, hash.clone());
+
+            // And close proposals
+            Self::close_proposal(hash);
+        }
+
+        Ok(())
+    }
+
+    /// Summarize voting and create referendums if proposals meet or exceed quorum threshold
+    fn tally_votes(index: MipsIndex, proposal_hash: T::Hash) {
+        if let Some(voting) = <Voting<T>>::get(proposal_hash) {
+            let aye_stake = voting
+                .ayes
+                .iter()
+                .fold(<BalanceOf<T>>::zero(), |acc, ayes| acc + ayes.1);
+
+            let nay_stake = voting
+                .nays
+                .iter()
+                .fold(<BalanceOf<T>>::zero(), |acc, nays| acc + nays.1);
+
+            let net_stake = aye_stake - nay_stake;
+
+            if net_stake >= Self::quorum_threshold() {
+                if let Some(proposal) = <Proposals<T>>::get(&proposal_hash) {
+                    sr_primitives::print("creating referendum");
+                    Self::create_referendum(index, proposal_hash.clone(), proposal);
+                }
+            }
+        }
+    }
+
+    /// Create a referendum object from a proposal.
+    /// Committee votes on this referendum instance
+    fn create_referendum(index: MipsIndex, proposal_hash: T::Hash, proposal: T::Proposal) {
+        <ReferendumMetadata<T>>::mutate(|metadata| metadata.push((index, proposal_hash)));
+        <Referendums<T>>::insert(proposal_hash.clone(), proposal);
+        Self::deposit_event(RawEvent::ReferendumCreated(index, proposal_hash.clone()));
+    }
+
+    /// Close a proposal. Voting ceases and proposal is removed from storage.
+    /// All deposits are unlocked and returned to respective stakers.
+    fn close_proposal(proposal_hash: T::Hash) {
+        if let Some(_) = <Voting<T>>::get(proposal_hash) {
```

Comment:
Just a comment (not need to be changed): If you are not going to use the returned value, you can use something like:
```
if <Voting<T>>::get( proposal_hahs).is_some() {
...
```